functions/auth.js

```diff
@@ -19,3 +30,164 @@ exports.passwordsMatch = (submittedPassword, actualPassword) =>
   submittedPassword.trim() === actualPassword.trim() ||
   lowercaseFirstChar(submittedPassword.trim()) ===
     lowercaseFirstChar(actualPassword.trim());
+
+/**
+ * Redirect the user to the authentication consent screen. The 'state' cookie is set for later state verification.
+ *
+ * @see https://github.com/lelylan/simple-oauth2/blob/3.x/API.md#authorizeurlauthorizeoptions--string
+ */
+// TODO: rename this authorize or similar?
+exports.connectI4AOAuth = functions.https.onRequest(async (req, res) => {
+  const { venueId } = req.query;
+
+  if (!AUTH_ORIGIN) {
+    throw new HttpsError("internal", "auth.origin is not configured properly");
+  }
+
+  if (!checkIfValidVenueId(venueId)) {
+    throw new HttpsError("invalid-argument", "venueId is not a valid venue id");
+  }
+
+  const authConfig = await fetchAuthConfig(venueId);
+
+  const authClient = createOAuth2Client(authConfig);
+
+  // TODO: configure this in cloud config and/or firestore or similar
+  const authCodeReturnUri = `${AUTH_ORIGIN}/auth/connect/i4a/handler?venueId=${venueId}`;
+
+  // TODO: configure this in cloud config and/or firestore or similar
+  const scope = "";
+
+  const redirectUri = authClient.authorizationCode.authorizeURL({
+    redirect_uri: authCodeReturnUri,
+    scope,
+  });
+
+  functions.logger.log("Redirecting to:", redirectUri);
+  res.redirect(redirectUri);
+});
+
+/**
+ * Exchanges a given auth code passed in the 'code' URL query parameter for a Firebase auth token.
+ * The request also needs to specify a 'state' query parameter which will be checked against the 'state' cookie.
+ */
+exports.connectI4AOAuthHandler = functions.https.onRequest(async (req, res) => {
+  const { venueId } = req.query;
+
+  if (!AUTH_ORIGIN) {
+    throw new HttpsError("internal", "auth.origin is not configured properly");
+  }
+
+  if (!checkIfValidVenueId(venueId)) {
+    throw new HttpsError("invalid-argument", "venueId is not a valid venue id");
+  }
+
+  const authConfig = await fetchAuthConfig(venueId);
+
+  const {
+    i4aApiKey,
+    i4aOAuthUserInfoUrl,
+    i4aGetUserMeetingInfoUrl,
+  } = authConfig;
+
+  const authClient = createOAuth2Client(authConfig);
+
+  const { code: authCode } = req.query;
+
+  functions.logger.log("Received auth code:", authCode);
+  const results = await authClient.authorizationCode.getToken({
+    code: authCode,
+    redirect_uri: "http://localhost:3000/in/foo", // TODO: what do we want to use here?
+  });
+  functions.logger.log("Auth code exchange result received:", results);
+
+  const { access_token: accessToken } = results;
+
+  // Retrieve the user's I4A User ID
+  const { id: i4aUserId } = await getJson(i4aOAuthUserInfoUrl, {
+    Authorization: `Bearer ${accessToken}`,
+  });
+
+  functions.logger.log("I4A User ID:", i4aUserId);
+
+  if (!i4aUserId) {
+    // TODO: redirect to some kind of '500 error' page?
+    res.redirect("/in/TODO/error");
+    return;
+  }
```

Comment:
[TODO] redirect to some kind of '500 error' page? Or maybe just throw a 500 error? This shouldn't really happen in reality, so it's probably fine to not show the user a perfectly pretty error page if it does.

Reply:
fixed in https://github.com/sparkletown/sparkle/pull/1523/commits/93a18c8a51b1c23ece18542ecbe856fcd65e6c67